tests/lazyflowRequestBuffer_test.py

```diff
@@ -0,0 +1,421 @@
+
+    Separate fixture to avoid unpacking in every test.
+    """
+    return TimeoutRaisingEvent()
+
+
+def install_finish_even(
+    obj: LazyflowRequestBuffer, func: str, event: Optional[TimeoutRaisingEvent] = None, n_calls: int = 1
+):
+    """
+    Helper function that wraps any function and sets an event after the function
+    has been called `n_calls` times.
+
+    Needed to ensure asynchronous calls are waited on. Not possible to wrap in a
+    mock - no guarantee the calls have been done when checking the mock.
+    """
+    old_func = getattr(obj, func)
+
+    n_calls_made = 0
+
+    event = event or TimeoutRaisingEvent()
+
+    def new_func(*args, **kwargs):
+        nonlocal n_calls_made
+        ret = old_func(*args, **kwargs)
+        n_calls_made += 1
+        if n_calls_made == n_calls:
+            event.set()
+        return ret
+
+    setattr(obj, func, new_func)
+
+    return event
+
+
+@pytest.fixture
+def default_waiting_func():
+    """Default fixture to use in request_buffer
+
+    Separate fixture to avoid unpacking
+    """
+    return WaitingFunc()
+
+
+@pytest.fixture
+def request_buffer(finished_evt: TimeoutRaisingEvent, default_context: Context, default_waiting_func: WaitingFunc):
+    """Default request buffer
+
+    Already preloaded with one running WaitingFunc task with the default context.
+
+    Hint use `default_waiting_func` fixture in tests to interact with this blocking task.
+    """
+    buffer = LazyflowRequestBuffer(1)
+    _ = install_finish_even(buffer, "decr", event=finished_evt)
+    buffer.submit(
+        default_waiting_func,
+        priority=default_context.priority,
+        viewport_ref=default_context.view_port,
+        stack_id=default_context.stack_id,
+        tile_no=default_context.tile_no,
+    )
+    default_waiting_func.started.wait_raise()
+    assert default_waiting_func.running
+    return buffer
+
+
+@pytest.mark.parametrize(
+    "n_concurrent_tasks",
+    [
+        -1,
+        -5,
+        0,
+    ],
+)
+def test_raises_on_init(n_concurrent_tasks: int):
+    with pytest.raises(RuntimeError):
+        _ = LazyflowRequestBuffer(n_concurrent_tasks)
+
+
+def test_task_calls_decr(
+    request_buffer: LazyflowRequestBuffer, finished_evt: TimeoutRaisingEvent, default_waiting_func: WaitingFunc
+):
+    assert request_buffer._active == 1  # pyright: ignore [reportPrivateUsage]
+
+    default_waiting_func.req_continue.set()
+    finished_evt.wait_raise()
+    assert not default_waiting_func.running
+    assert request_buffer._active == 0  # pyright: ignore [reportPrivateUsage]
+    assert request_buffer._failed == 0  # pyright: ignore [reportPrivateUsage]
+
+
+@pytest.mark.qt_no_exception_capture
+def test_task_calls_decr_on_error(default_context: Context):
+    # can't use default fixture - need to ensure patched function is called twice before emitting signal
+    request_buffer = buffer = LazyflowRequestBuffer(1)
+    finished_evt = install_finish_even(buffer, "decr", n_calls=1)
```

Comment:
```suggestion
    request_buffer = LazyflowRequestBuffer(1)
    finished_evt = install_finish_even(request_buffer, "decr", n_calls=1)
```

not sure what this is for